src/rust/src/x509.rs

```diff
@@ -43,6 +43,454 @@ lazy_static::lazy_static! {
+    #[implicit(1)]
+    _issuer_unique_id: Option<asn1::BitString<'a>>,
+    #[implicit(2)]
+    _subject_unique_id: Option<asn1::BitString<'a>>,
+    #[explicit(3)]
+    extensions: Option<Extensions<'a>>,
+}
+
+pub(crate) type Name<'a> = asn1::SequenceOf<'a, asn1::SetOf<'a, AttributeTypeValue<'a>>>;
+
+#[derive(asn1::Asn1Read, asn1::Asn1Write)]
+pub(crate) struct AttributeTypeValue<'a> {
+    pub(crate) type_id: asn1::ObjectIdentifier<'a>,
+    pub(crate) value: asn1::Tlv<'a>,
+}
+
+#[derive(asn1::Asn1Read, asn1::Asn1Write)]
+enum Asn1Time {
+    UtcTime(asn1::UtcTime),
+    GeneralizedTime(asn1::GeneralizedTime),
+}
+
+#[derive(asn1::Asn1Read, asn1::Asn1Write)]
+pub(crate) struct Validity {
+    not_before: Asn1Time,
+    not_after: Asn1Time,
+}
+
+#[ouroboros::self_referencing]
+struct OwnedRawCertificate {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    value: RawCertificate<'this>,
+}
+
+#[pyo3::prelude::pyclass]
+struct Certificate {
+    raw: OwnedRawCertificate,
+    cached_extensions: Option<pyo3::PyObject>,
+}
+
+#[pyo3::prelude::pyproto]
+impl pyo3::class::basic::PyObjectProtocol for Certificate {
+    fn __hash__(&self) -> u64 {
+        let mut hasher = DefaultHasher::new();
+        self.raw.borrow_data().hash(&mut hasher);
+        hasher.finish()
+    }
+
+    fn __richcmp__(
+        &self,
+        other: pyo3::pycell::PyRef<Certificate>,
+        op: pyo3::class::basic::CompareOp,
+    ) -> pyo3::PyResult<bool> {
+        match op {
+            pyo3::class::basic::CompareOp::Eq => {
+                Ok(self.raw.borrow_data() == other.raw.borrow_data())
+            }
+            pyo3::class::basic::CompareOp::Ne => {
+                Ok(self.raw.borrow_data() != other.raw.borrow_data())
+            }
+            _ => Err(pyo3::exceptions::PyTypeError::new_err(
+                "Certificates cannot be ordered",
+            )),
+        }
+    }
+
+    fn __repr__(&self) -> pyo3::PyResult<String> {
+        let mut repr = String::from("<Certificate(subject=<Name(");
+        let mut vec = Vec::new();
+        for rdn in self.raw.borrow_value().tbs_cert.subject.clone() {
+            for attribute in rdn {
+                let mut attr = attribute.type_id.to_string();
+                attr.push('=');
+                attr.push_str(std::str::from_utf8(attribute.value.data()).unwrap());
+                vec.push(attr);
+            }
+        }
+        repr.push_str(&vec.join(", "));
+        repr.push_str(")>, ...)>");
+        Ok(repr)
+    }
+}
+
+#[pyo3::prelude::pymethods]
+impl Certificate {
+    fn __deepcopy__(
+        slf: pyo3::pycell::PyRef<'_, Self>,
+        _memo: pyo3::PyObject,
+    ) -> pyo3::pycell::PyRef<'_, Self> {
+        slf
+    }
+
+    fn public_key<'p>(&self, py: pyo3::Python<'p>) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        let serialized = pyo3::types::PyBytes::new(
+            py,
+            &asn1::write_single(&self.raw.borrow_value().tbs_cert.spki),
```

Comment:
Can you put a TODO comment on here admonishing us that this performs a copy that shouldn't be required?